PhysicsTools/PatAlgos/src/MuonMvaEstimator.cc

```diff
@@ -0,0 +1,139 @@
+#include "DataFormats/MuonReco/interface/MuonSelectors.h"
+#include "DataFormats/VertexReco/interface/Vertex.h"
+#include "DataFormats/JetReco/interface/PFJet.h"
+#include "DataFormats/JetReco/interface/PFJetCollection.h"
+#include "DataFormats/Candidate/interface/Candidate.h"
+#include "JetMETCorrections/JetCorrector/interface/JetCorrector.h"
+
+using namespace pat;
+
+MuonMvaEstimator::MuonMvaEstimator():
+  tmvaReader_("!Color:!Silent:Error"),
+  initialized_(false),
+  mva_(0),
+  dRmax_(0)
+{}
+
+void MuonMvaEstimator::initialize(std::string weightsfile,
+				  float dRmax)
+{ 
+  if (initialized_) return;
+  tmvaReader_.AddVariable("LepGood_pt",                    &pt_               );
+  tmvaReader_.AddVariable("LepGood_eta",                   &eta_              );
+  tmvaReader_.AddVariable("LepGood_jetNDauChargedMVASel",  &jetNDauCharged_   );
+  tmvaReader_.AddVariable("LepGood_miniRelIsoCharged",     &miniRelIsoCharged_);
+  tmvaReader_.AddVariable("LepGood_miniRelIsoNeutral",     &miniRelIsoNeutral_);
+  tmvaReader_.AddVariable("LepGood_jetPtRelv2",            &jetPtRel_         );
+  tmvaReader_.AddVariable("min(LepGood_jetPtRatiov2,1.5)", &jetPtRatio_       );
+  tmvaReader_.AddVariable("max(LepGood_jetBTagCSV,0)",     &jetBTagCSV_       );
+  tmvaReader_.AddVariable("LepGood_sip3d",                 &sip_              );
+  tmvaReader_.AddVariable("log(abs(LepGood_dxy))",         &log_abs_dxyBS_    ); 
+  tmvaReader_.AddVariable("log(abs(LepGood_dz))",          &log_abs_dzPV_     );
+  tmvaReader_.AddVariable("LepGood_segmentCompatibility",  &segmentCompatibility_);
+  tmvaReader_.BookMVA("BDTG",weightsfile);
+  dRmax_ = dRmax;
+  initialized_ = true;
+};
+
+float ptRel(const reco::Candidate::LorentzVector& muP4, 
+	    const reco::Candidate::LorentzVector& jetP4, 
+	    bool subtractMuon=true) 
+{
+  reco::Candidate::LorentzVector jp4 = jetP4;
+  if (subtractMuon) jp4-=muP4;
+  float dot = muP4.Vect().Dot( jp4.Vect() );
+  float ptrel = muP4.P2() - dot*dot/jp4.P2();
+  ptrel = ptrel>0 ? sqrt(ptrel) : 0.0;
+  return ptrel;
+}
+
+void MuonMvaEstimator::computeMva(const pat::Muon& muon,
+				  const reco::Vertex& vertex,
+				  const reco::JetTagCollection& bTags,
+				  const reco::JetCorrector* correctorL1,
+				  const reco::JetCorrector* correctorL1L2L3Res)
+{
+  if (not initialized_) 
+    throw cms::Exception("FatalError") << "MuonMVA is not initialized";
+  pt_                   = muon.pt();
+  eta_                  = muon.eta();
+  segmentCompatibility_ = muon.segmentCompatibility();
+  miniRelIsoCharged_ = muon.miniPFIsolation().chargedHadronIso();
+  miniRelIsoNeutral_ = muon.miniPFIsolation().neutralHadronIso();
+
+  double dB2D  = muon.dB(pat::Muon::BS2D);
+  double dB3D  = muon.dB(pat::Muon::PV3D);
+  double edB3D = muon.edB(pat::Muon::PV3D);
+  sip_  = edB3D>0?fabs(dB3D/edB3D):0.0; 
+  log_abs_dxyBS_     = log(fabs(dB2D)); 
+  log_abs_dzPV_      = log(fabs(muon.muonBestTrack()->dz(vertex.position())));
```

Comment:
log inputs should be protected from zero values to avoid NaNs

Reply:
.. perhaps a normal result in this context is never exactly zero; m/be OK then

Reply:
fixed